### PR TITLE
Remove `ty_to_def_id`

### LINKED ISSUE #52341: `Ty::ty_to_def_id` does not yield all possible `DefId`s that it could
I don't know if this is really problem, but e.g. the `ReachEverythingInTheInterfaceVisitor` reimplements this function (and yields a bunch of other `DefId`s, too.

This needs some looking at

Mostly leaving this here so I don't forget

### PR DESCRIPTION
**Pull request by oli-obk**
fixes https://github.com/rust-lang/rust/issues/52341

The uses were mostly convenience and generally "too powerful" (would also have worked for types that weren't interesting at the use site)

r? @eddyb

### CODE REVIEW

File: src/librustc/middle/dead.rs [hunk 1]
```diff
@@ -469,13 +469,9 @@ impl<'a, 'tcx> DeadVisitor<'a, 'tcx> {
...
     fn should_warn_about_field(&mut self, field: &hir::StructField) -> bool {
         let field_type = self.tcx.type_of(self.tcx.hir.local_def_id(field.id));
-        let is_marker_field = match field_type.ty_to_def_id() {
-            Some(def_id) => self.tcx.lang_items().items().iter().any(|item| *item == Some(def_id)),
-            _ => false
-        };
         !field.is_positional()
             && !self.symbol_is_live(field.id, None)
-            && !is_marker_field
+            && !field_type.is_phantom_data()
```
Comment (eddyb) [inline src/librustc/middle/dead.rs]: Heh, are the only two lang item types, `PhantomData` and `Box`?

Comment (oli-obk) [reply src/librustc/middle/dead.rs]: Ugh. I don't see why Box should be special. Depending on the `lang_item`nes of a type seems rather fragile.

File: src/librustc_codegen_llvm/back/write.rs [hunk 2]
```diff
@@ -399,6 +399,9 @@ impl CodegenContext {
 }
 
 struct DiagnosticHandlers<'a> {
+    #[allow(dead_code)]
+    // This value is not actually dead, llcx has pointers to it and needs these pointers to be alive
+    // until Drop is executed on this object
     inner: Box<(&'a CodegenContext, &'a Handler)>,
```
Comment (eddyb) [inline src/librustc_codegen_llvm/back/write.rs]: Do you mean the heap allocation itself? This is really confusing, maybe a raw pointer should be kept instead, with a manual `Drop` impl that uses `Box::from_raw`.

File: src/librustc/middle/dead.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc/middle/dead.rs]: I agree it shouldn't be, but once upon a time there were many lang items, one for each of "{co,contra,in}variant {lifetime,type}" (so 6 at least), so maybe that's where it's leftover from.
I'm just amused only two types are left and the `Box` case actually hides some warnings!

File: src/librustc_codegen_llvm/back/write.rs (hunk 2 above)
Comment (eddyb) [reply src/librustc_codegen_llvm/back/write.rs]: Also maybe the comment should be phrased to be about the callback having a copy of the pointer or something. It's not really "self-referential" insofar as much as it is scoping a resource for the duration C++ needs it.

**Pull request merged**